---
title: "Building and Using Ensembl-based Annotation Packages with ensembldb"
author: "Johannes Rainer"
date: "June 26, 2016"
output:
  ioslides_presentation:
    css: style.css
    widescreen: false
---
<style scoped>
ul > li:before {
    color: #1a81c2;
    vertical-align: middle;
    font-family: "Arial Black";
    font-weight: 900;
    margin-left: -.85em;
}
</style>

# Building and Using Ensembl-based Annotation Packages with `ensembldb`

**Johannes Rainer** (EURAC), BioC 2016 Stanford

Clone me: <http://github.com/jotsetung/Bioc2016-ensembldb>.

## Introduction

* `TxDb` objects from `GenomicFeatures` provide gene model annotations.
* `ensembldb` package defines `EnsDb` class also providing gene models.
* `EnsDB` object/database:
    +   Designed for Ensembl annotations: new attributes *gene biotype* and *tx biotype.*
    +   Allows to query specific annotations using a simple **filter framework**.
    +   Methods allow to specify the type of the returned object.
    +   Interchangeable with `TxDb` objects.
    +   Support `AnnotationDbi`'s `select` method.


## Usage: Query gene, transcript, exon information

-   `EnsDb` objects can be used just like `TxDb` objects.
-   Available methods:
    -   `genes`
    -   `transcripts`
    -   `transcriptsBy`
    -   `exons`
    -   `exonsBy`
    -   `cdsBy`
    -   `fiveUTRsByTranscripts`
    -   `threeUTRsByTranscripts`


----

**Example**: get gene and transcript data

```{r message=FALSE}
## Load an EnsDb package matching Ensembl version 81
library(EnsDb.Hsapiens.v81)
edb <- EnsDb.Hsapiens.v81

## Now just get all genes
genes(edb)
```

----

**Example**: get gene and transcript data

```{r message=FALSE}
## Alternatively, we can specify to return the results as DataFrame
transcripts(edb, return.type="DataFrame")
```

----

**Example**: Use filters to fetch specific information.

```{r }
####
## Task: retrieve genes encoded on chromosome Y.
## Create a filter object
sf <- SeqnameFilter("Y")

## Retrieve the data
genes(edb, filter=sf)
```

## Supported filters

* For genes:
    +   `GeneidFilter`
    +   `GenenameFilter`
    +   `EntrezidFilter`
    +   `GenebiotypeFilter`
* For transcripts:
    +   `TxidFilter`
    +   `TxbiotypeFilter`
* For exons:
    +   `ExonidFilter`
    +   `ExonrankFilter`

## Supported filters

- _Generic_ filters:
    -   `SeqnameFilter`
    -   `SeqstrandFilter`
    -   `SeqstartFilter`
    -   `SeqendFilter`
    -   `GRangesFilter`: condition can be _within_ or _overlapping_.

- Multiple filters can be combined with a logical _AND_.
- Each filter supports 1:n values and also a /like/ condition.

----

**Example**: combine filters.

```{r }
##  Example for a GRangesFilter; condition could be "within" or "overlapping"
grf <- GRangesFilter(GRanges(17, IRanges(59000000, 59200000)),
		     condition="within")

## Combine filters: add a GenebiotypeFilter
## to get all genes in the region except pre-miRNAs and snRNAs.
genes(edb, filter=list(grf,
		       GenebiotypeFilter(c("miRNA", "snRNA"),
					 condition="!=")))
```

## Annotation for feature counting

-   `EnsDb` databases can also provide gene model information for feature counting.
-   In combination with `summarizeOverlaps` from the `GenomicAlignments` package.
-   __Example__: get gene models for feature counting with `summarizeOverlaps`.

```{r eval=FALSE}
library(GenomicAlignments)
library(BiocParallel)
## Get exons by gene, for chromosomes 1:22, X, Y, excluding locus reference genomic genes (LRG)
exns <- exonsBy(edb, by="gene", filter=list(SeqnameFilter(c(1:22, "X", "Y")),
					    GeneidFilter("ENSG%", "like")))

bfl <- BamFileList(dir("data/bam", pattern=".bam$", full.names=TRUE),
		   asMates=TRUE, yieldSize=1e+6, obeyQname=TRUE)
## Define a ScanBamParam with a mapping quality filter.
sbp <- ScanBamParam(mapqFilter=30)

## Do the gene counting
geneCounts <- bplapply(bfl, FUN=summarizeOverlaps, features=exns,
		       mode="IntersectionStrict", ignore.strand=TRUE,
		       singleEnd=FALSE, fragments=TRUE, param=sbp)
geneCounts <- do.call(cbind, geneCounts)
```

----

-   In combination with `Rsubread`'s `featureCount` function.
-   __Example__: gene models for `featureCount`.

```{r eval=FALSE}
## Convert the exon list to SAF format
saf <- toSAF(exns)

####
##  Do the feature counting using the Rsubread package
library(Rsubread)
bamf <- dir("data/bam", pattern=".bam$", full.names=TRUE)
cnts <- featureCounts(files=bamf, annot.ext=saf, isPairedEnd=TRUE, nthreads=1)
```

## Get genomic or transcript sequences

-   `BSGenome` packages provide genomic sequence for a variety of species.
-   Most are based on UCSC data, that use different genome
    release names than Ensembl (e.g. *hg19* instead of *GRCh37*) and have a different
    chromosome naming style.
-   __Example__: get mRNA sequences for a gene's transcripts.

```{r message=FALSE, results="hold"}
##  Get the genome version of the package:
unique(genome(edb))
ensemblVersion(edb)

## Where to get this genome sequence? -> AnnotationHub
library(AnnotationHub)
ah <- AnnotationHub()
## Search for all human Ensembl files for release 81.
query(ah, pattern=c("Ensembl", "homo sapiens", "81"))

fafile <- ah[["AH49186"]]

```

----

- __Example__: get mRNA sequences for a gene's transcripts.

```{r message=FALSE}
## Easier way: use the dedicated getGenomeFaFile method
fafile <- getGenomeFaFile(edb)

## Task: get transcript sequences for SKA2
## Get all exons by transcript.
ska2Txs <- exonsBy(edb, by="tx", filter=GenenameFilter("SKA2"))

## Use extractTranscriptSeqs from GenomicFeatures; alternative: getSeq from Biostrings
txSeqs <- extractTranscriptSeqs(fafile, ska2Txs)
txSeqs
```

-   That's convenient, especially if we're using an `EnsDb` for a non-standard
    organism.

----

-   **But**: wouldn't it be nice to use `BSGenome` packages?
-   __Example__: use `EnsDb` with UCSC style chromosome names.

```{r message=FALSE}
library(BSgenome.Hsapiens.UCSC.hg38)

head(seqlevels(BSgenome.Hsapiens.UCSC.hg38))
head(seqlevels(edb))
## Doesn't work, because of different chromosome naming.

## Solution: change the seqlevel style
seqlevelsStyle(edb) <- "UCSC"

head(seqlevels(edb))
genes(edb, filter=SeqnameFilter("chrY"))

```

----

-   __Example__: use `EnsDb` with UCSC style chromosome names.

```{r message=FALSE}
## Get the exons by gene for SKA2
ska2Txs <- exonsBy(edb, filter=GenenameFilter("SKA2"))

## Get the tx sequence using the BSgenome package.
extractTranscriptSeqs(BSgenome.Hsapiens.UCSC.hg38, ska2Txs)

## Compare to the FaFile from Ensembl.
txSeqs

## Change back, so we don't break following examples.
seqlevelsStyle(edb) <- "Ensembl"
```
